packages/react/roving-focus/src/RovingFocusGroup.tsx

```diff
@@ -1,144 +1,237 @@
+
 const ITEM_NAME = 'RovingFocusItem';
+const ITEM_DEFAULT_TAG = 'span';
 
-type UseRovingFocusItemOptions = { disabled?: boolean; active?: boolean };
+type RovingFocusItemOwnProps = Polymorphic.Merge<
+  Polymorphic.OwnProps<typeof Primitive>,
+  {
+    focusable?: boolean;
+    active?: boolean;
+  }
+>;
 
-function useRovingFocus({ disabled, active = false }: UseRovingFocusItemOptions) {
-  const itemId = useId();
+type RovingFocusItemPrimitive = Polymorphic.ForwardRefComponent<
+  typeof ITEM_DEFAULT_TAG,
+  RovingFocusItemOwnProps
+>;
+
+const RovingFocusItem = React.forwardRef((props, forwardedRef) => {
+  const { as = ITEM_DEFAULT_TAG, focusable = true, active = false, ...itemProps } = props;
+  const id = useId();
+  const ref = React.useRef<HTMLElement>(null);
+  const composedRefs = useComposedRefs(forwardedRef, ref);
   const context = useRovingFocusContext(ITEM_NAME);
-  const isTabStop = itemId === context.tabStopId;
+  const isCurrentTabStop = context.currentTabStopId === id;
+  const isCurrentTabStopRef = React.useRef(isCurrentTabStop);
+  React.useEffect(() => void (isCurrentTabStopRef.current = isCurrentTabStop));
 
+  // We keep an up to date map of every item. We do this on every render
+  // to make sure the map insertion order reflects the DOM order.
   React.useEffect(() => {
-    context.itemMap.set(itemId, active);
-    return () => {
-      context.itemMap.delete(itemId);
-    };
-  }, [active, itemId, context.itemMap]);
-
-  if (disabled) {
-    return {
-      tabIndex: -1,
-      // we prevent focusing disabled items on `mousedown`. Even though the item has tabIndex={-1},
-      // that only means take it out of the tabbable order
-      onMouseDown: (event: React.MouseEvent) => event.preventDefault(),
-    };
-  }
+    context.itemMap.set(id, { ref, id, focusable, active });
+    return () => void context.itemMap.delete(id);
+  });
+
+  return (
+    <Primitive
+      tabIndex={isCurrentTabStop ? 0 : -1}
+      data-orientation={context.orientation}
+      {...itemProps}
+      as={as}
+      ref={composedRefs}
+      onMouseDown={composeEventHandlers(props.onMouseDown, (event) => {
+        // We prevent focusing non-focusable items on `mousedown`.
+        // Even though the item has tabIndex={-1}, that only means take it out of the tab order.
+        if (!focusable) event.preventDefault();
+        else context.onInteract();
+      })}
+      onFocus={composeEventHandlers(props.onFocus, () => context.onItemFocus(id))}
+      onKeyDown={composeEventHandlers(props.onKeyDown, (event) => {
+        if (event.key === 'Tab' && event.shiftKey) {
+          context.onItemShiftTab();
+          return;
+        }
+
+        if (event.target !== event.currentTarget) return;
+
+        // stop key events from propagating to parent in case we're in a nested roving focus group
+        if (KEYS.includes(event.key)) event.stopPropagation();
```

Comment:
I have a case where I'd like to handle `keydown` in a parent element rather than binding directly to the roving items, has this come up before? wondering whether it's worth having an option for this.

Reply:
You mean perhaps it's not a good idea to stop propagation early here?

This made me realise that it's not needed anymore actually because I now check on the line above that it's an even happening on the current target, and not a bubbling event.

So I can remove it, and I think that solves your question too then?

Reply:
Yes exactly, this solves it.